### ISSUE TITLE
TLS version change

### ISSUE THREAD

**Issue by SaraAlizadeh0882** [timothy-wong/ukgasapi#2]
Hi Timothy,

I'm using ukgasapi package to pull the data from MIPI. A recent change to National Grid internet policy means that it blocks anything less than TLS version 1.2. So, after this change the package doesn't work. I appreciate if you could please force a change to TLS 1.2 and then I think it should work as before without any further issue.
I look forward to hearing from you.

Kind regards,

Sara Alizadeh

**Comment by timothywong731** [comment]
Actually would you please try this for me? This is the updated function which you can pass additional arguments to curlPerfrom(). Try pass the .opt argument to dataItemExplorer() function through the '...' at the end.

```
dataItemExplorer<- function(dataitems,
                            fromdate,
                            todate,
                            datetype="gasday",
                            latestflag="Y",
                            applicableforflag="Y",
                            batchsize = -1,
                            apiurl = paste0("https://marketinformation.natgrid.co.uk/",
                                            "MIPIws-public/public/publicwebservice.asmx"),
                            ...) {

  # Using batch mode
  if (batchsize > 0) {

    # Calculate batch IDs
    batch.quotient <- as.integer(as.Date(todate) - as.Date(fromdate) + 1) %/% batchsize
    batch.modulo <- as.integer(as.Date(todate) - as.Date(fromdate) + 1) %% batchsize

    # Create a row for each batch
    if (batch.quotient >0) {
      list.of.dates <- base::data.frame(
        start = 1:batch.quotient * batchsize - batchsize+1,
        end = 1:batch.quotient * batchsize)
    } else {
      list.of.dates <- base::data.frame(start = integer(),
                                        end = integer())
    }

    # Append a row if there's any remainder (modulo)
    if(batch.modulo > 0) {
      list.of.dates <- rbind(list.of.dates,
                             base::data.frame(
                               start = batch.quotient * batchsize+1,
                               end = batch.quotient * batchsize + batch.modulo))
    }

    # Create a sequence of all dates
    all.dates <- seq.Date(as.Date(fromdate), as.Date(todate),1)

    # Create an empty list of data frames
    results <- list()

    # Loop through all batch an run the API
    for (i in 1:nrow(list.of.dates)) {
      batch.fromdate <- all.dates[list.of.dates[i,]$start]
      batch.todate <- all.dates[list.of.dates[i,]$end]
      results[[i]] <- dataItemExplorer(
        dataitems = dataitems,
        fromdate = batch.fromdate,
        todate = batch.todate,
        datetype = datetype,
        latestflag = latestflag,
        applicableforflag = applicableforflag,
        batchsize = -1,
        apiurl = apiurl)
    }

    return(do.call(rbind, results))

  } else {
    # Non-batch mode

    # Creates SOAP XML request
    soap.request <- base::paste0('<?xml version="1.0" encoding="utf-8"?><soap12:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope"><soap12:Body><GetPublicationDataWM xmlns="http://www.NationalGrid.com/MIPI/"><reqObject><LatestFlag>',latestflag,'</LatestFlag><ApplicableForFlag>',applicableforflag,'</ApplicableForFlag><ToDate>',todate,'</ToDate><FromDate>',fromdate,'</FromDate><DateType>',datetype,'</DateType><PublicationObjectNameList>',paste0('<string>',dataitems,'</string>', collapse = ''),'</PublicationObjectNameList></reqObject></GetPublicationDataWM></soap12:Body></soap12:Envelope>')

    # Initialises a text gatherer
    h <- RCurl::basicTextGatherer()
    h$reset()

    # Initiates SOAP request via HTTP
    RCurl::curlPerform(url = apiurl,
                       httpheader = c(Accept="text/xml", 'Content-Type'='application/soap+xml; charset=utf-8'),
                       postfields = soap.request,
                       verbose=TRUE,
                       writefunction = h$update,
                       ...)

    # Writes SOAP response into character
    soap.response <- h$value()

    # converts SOAP response into an XML object
    soap.response.doc <- XML::xmlTreeParse(soap.response, replaceEntities = TRUE , useInternalNodes = TRUE)

    # converts SOAP response into a list of objects
    objects <- (XML::xmlToList(soap.response.doc,simplify = TRUE))[[1]][[1]]

    list.of.data.frames<-list()

    for(i in 1:length(objects)) {
      # Porcessing data frames
      rows <- objects[[i]][["PublicationObjectData"]]

      list.of.rows <- base::data.frame()

      for(r in 1:length(rows)){
        # Processing individual row
        list.of.rows <- base::rbind(list.of.rows,as.data.frame(rows[r][[1]],stringsAsFactors = FALSE))
      }
      list.of.rows$PublicationObjectName <- objects[[i]][[1]]

      list.of.data.frames <- base::rbind(list.of.data.frames, as.data.frame(list.of.rows))
    }

    # Convert all columns to appropiate types before returning data frame
    result <- base::within(data = list.of.data.frames,expr = {
      ApplicableAt <- strptime(ApplicableAt,"%Y-%m-%dT%H:%M:%SZ","UTC")
      ApplicableFor <- strptime(ApplicableFor,"%Y-%m-%dT%H:%M:%SZ","UTC")
      Value <- as.numeric(Value)
      GeneratedTimeStamp <- strptime(GeneratedTimeStamp,"%Y-%m-%dT%H:%M:%SZ","UTC")
      QualityIndicator <- factor(QualityIndicator)
      Substituted <- factor(QualityIndicator)
      CreatedDate <- strptime(CreatedDate,"%Y-%m-%dT%H:%M:%SZ","UTC")
      PublicationObjectName <- as.factor(PublicationObjectName)
    })

    # Prefer using POSIXct rather than POSIXlt
    result$ApplicableAt <- as.POSIXct(result$ApplicableAt)
    result$ApplicableFor <- as.POSIXct(result$ApplicableFor)
    result$GeneratedTimeStamp <- as.POSIXct(result$GeneratedTimeStamp)
    result$CreatedDate <- as.POSIXct(result$CreatedDate)

    return (result)
  }
}
```

I followed this page on [stackoverflow](https://stackoverflow.com/questions/38130293/tls-v1-1-tls-v1-2-support-in-rcurl)

If it works, I can push it to CRAN. Please let me know.

**Comment by timothywong731** [comment]
SSL problem fixed in #3 - **v0.18 is on the way to CRAN**
